stage0/gc.go

```diff
@@ -65,55 +66,94 @@ type mount struct {
 	id         int
 	parentId   int
 	mountPoint string
+	mode       string
 }
 
 type mounts []*mount
 
-// Less ensures that mounts are sorted in an order we can unmount; child before
-// parent. This works by pushing all child mounts up to the front. Parent
-// mounts are halted once they encounter a child. The actual sequential order
-// of the mounts doesn't matter; only that children are in front of parents.
-func (m mounts) Less(i, j int) bool { return m[i].id != m[j].parentId }
-func (m mounts) Len() int           { return len(m) }
-func (m mounts) Swap(i, j int)      { m[i], m[j] = m[j], m[i] }
+// getMountDepth searches the list for all ancestors of the mount at index i
+func (m mounts) getMountDepth(i int) int {
+	ancestorMap := make(map[int]struct{}, 0)
+	current := m[i]
+	for found := true; found; {
+		found = false
+		for _, mnt := range m {
+			if mnt.id == current.parentId {
+				ancestorMap[mnt.id] = struct{}{}
+				current = mnt
+				found = true
+				break
+			}
+		}
+	}
+	return len(ancestorMap)
+}
+
+// Less ensures that mounts are sorted in an order we can unmount; descendant before ancestor.
+// The requirement of transitivity for Less has to be fulfilled otherwise the sort algorithm will fail.
+func (m mounts) Less(i, j int) (result bool) { return m.getMountDepth(i) >= m.getMountDepth(j) }
```

Comment:
I think `>` instead of `>=` would work and do less permutations in the sort: when the two elements are equals, `Less` could return `false` because they don't need to be permuted.